### PR TITLE
feat: Support for LBExternalAddresses annotation in service conversion

### LINKED ISSUE #43656: Allow overriding external IP of multi-cluster east-west gateways
(This is used to request new product features, please visit <https://discuss.istio.io> for questions on using Istio)

**Describe the feature request**
I'm using a multi-primary multi-network Istio setup across 3 clusters, and 2 of those clusters are behind NAT. Those 2 clusters use MetalLB to get loadbalancer services, which works great. The issue is that due to MetalLB setting the status of those loadbalancers to an internal IP (which is correct), Istio then tries to use that IP to route traffic to the other clusters east-west gateway.

**Describe alternatives you've considered**

After digging around in the documentation and in the code for a bit, I saw that Istio also respects the `externalIPs` attribute of services, however the IPs configured there are added to the ones in the status, which would result in about half of the requests failing. 

There are two solutions for this (at least that I came up with)
- taking the value of externalIPs over the status IP, which would allow specifying a gateway's public IP via that field, allthough this would be a backwards incompatible change
- Adding an optional annotation to the LoadBalancer service, which would allow overriding the address via an annotation, this has the advantage that it would also work with DNS-based addresses and not just raw IPs

The latter would be imo the better choice, and to test it I've created a small patch locally, but I thought it might be a good idea to open an issue for discussion before creating a PR.

```diff
diff --git a/pilot/pkg/serviceregistry/kube/conversion.go b/pilot/pkg/serviceregistry/kube/conversion.go
index 075d62d171..bd383d3fd7 100644
--- a/pilot/pkg/serviceregistry/kube/conversion.go
+++ b/pilot/pkg/serviceregistry/kube/conversion.go
@@ -42,6 +42,8 @@ const (
        // It is used for multi-cluster scenario, and with nodePort type gateway service.
        // TODO: move to API
        NodeSelectorAnnotation = "traffic.istio.io/nodeSelector"
+
+       ExternalAddressAnnotation = "traffic.istio.io/externalAddress"
 )

 func convertPort(port corev1.ServicePort) *model.Port {
@@ -167,6 +169,9 @@ func ConvertService(svc corev1.Service, domainSuffix string, clusterID cluster.I
        istioService.Attributes.NodeLocal = nodeLocal
        istioService.Attributes.ClusterExternalAddresses.AddAddressesFor(clusterID, svc.Spec.ExternalIPs)

+       if addr, ok := svc.Annotations[ExternalAddressAnnotation]; ok {
+               istioService.Attributes.ClusterExternalAddresses.SetAddressesFor(clusterID, []string{addr})
+       }
        return istioService
 }

```

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[x] Multi Cluster
[ ] Virtual Machine
[x] Multi Control Plane

**Additional context**

### PR DESCRIPTION
**Pull request by frauniki**
This PR adds the ability to override the external address from the annotation when converting a load balancer type service to an istio service.

In the current implementation, load balancer type services resolve external addresses using only the addresses set in `.status.loadBalancer.ingress` by the load balancer controller, etc.
By adding this feature, istio users will be able to override the external address of the load balancer with a customized ip address or dns record, if necessary.

Adding this feature can support cases where for some reason you want to connect to a service's load balancer using customized DNS records, etc.
For example, the DNS records automatically generated by the load balancer controller are set to a private zone and cannot be resolved directly from other networks.

- Fixed #43656

This is my first commit to this repository, so please let me know if I am doing something wrong!

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by stevenctl** [comment]
This LGTM. There is another stream of work (#45342) that enables something ~similar, but not this behavior exactly. 

It's sort of adding another API, so I'd like one other contributor to approve as well. cc @hzxuzhonghu  or @howardjohn 

Thanks for the contribution.

File: pilot/pkg/serviceregistry/kube/conversion.go [hunk 1]
```diff
@@ -42,6 +42,13 @@ const (
 	// It is used for multi-cluster scenario, and with nodePort type gateway service.
 	// TODO: move to API
 	NodeSelectorAnnotation = "traffic.istio.io/nodeSelector"
+
+	// LBExternalAddressesAnnotation overrides the external address value with a LoadBalancer type service.
```
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/conversion.go:46]: Somewhere in the doc/comment, make sure we emphasize this overwrites and does not append.

Comment (frauniki) [reply pilot/pkg/serviceregistry/kube/conversion.go]: Yes, add the same annotations and documentation to [istio/api](https://github.com/istio/api) once this PR is merged.
https://github.com/istio/api/blob/master/annotation/annotations.yaml

Comment (frauniki) [reply pilot/pkg/serviceregistry/kube/conversion.go]: https://github.com/istio/api/pull/2858

**Comment by hzxuzhonghu** [comment]
I donot understand the original issue actually. We support getting network address from status.Ingress and spec.ExternalIPs. And i know metalb set the svc.status.Ingress.

**Comment by frauniki** [comment]
First, when configuring an Istio Multi Cluster, assume that each cluster is in a private network behind a different NAT or in a different VPC.
For example, if you create an istio east-west gateway service for one of the clusters with the LoadBalancer type, the service's `.spec.ExternalIPs` may be set to private ip.
The case commented in #43656 is a case using MetalLB, but this problem may also occur in public clouds such as AWS, depending on the settings of the load balancer controller, etc. being used. (Because Private Load Balancer can be used in case of AWS)

In such a case, istio sets the external address to an ip address that can only be accessed from the same private network.
However, since the configured external address can only be accessed within the same NAT, another istio cluster built on a different NAT or VPC will not be able to connect.

To solve this problem, it is necessary to configure the target east-west gateway service with a public external address that can directly receive inbound traffic from outside the NAT. (For example, a public ip used in NAT or an EIP in AWS).
This requires an annotation that overrides the service's external address.

istio

**Comment by hzxuzhonghu** [comment]
OK, this case is currently not handled by k8s service no matter what svc type. But from the very begining we have a meshNetworks that allow user to specify any address/domain.  I think it can help

**Comment by frauniki** [comment]
Perhaps you are referring to the (MeshNetwork)[https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshNetworks] in the global mesh option, but yes, it may be possible to solve this problem by manually configuring all mesh configurations using the MeshNetwork.
However, if I use MeshNetwork in this configuration, I have to manually declare and manage all the clusters that make up the mesh for each cluster, and cluster discovery by istio will not work at all.
I am considering using MeshNetwork as a last resort, but I think it would be even better if cluster discovery using service could support these functions:)

**Comment by hzxuzhonghu** [comment]
We have supported so many kind of network gateway discovery ways. It is quite confusing to endusers, I think in the long run we not adding one after one as a separate point,. I believe maybe we should design a new CRD. It should cover service of LB type, NodePort, and external Address.

**Comment by frauniki** [comment]
> I believe maybe we should design a new CRD. It should cover service of LB type, NodePort, and external Address.

Do you mean implementing NetworkGateway as a CRD and using current services to support NetworkGateway discovery while also supporting NetworkGateway customization via CRD?
https://github.com/istio/istio/blob/master/pilot/pkg/model/network.go#L38-L47

For example, like this

```yaml
apiVersion: v1
kind: Service
metadata:
  name: eastwest-gateway
  labels:
    istio: eastwest-gateway
spec:
  type: LoadBalancer
  ports:
    - name: status-port
      port: 15021
      targetPort: 15021
    - name: tls
      port: 15443
      targetPort: 15443
---
apiVersion: networking.istio.io/v1alpha3
kind: NetworkGateway
metadata:
  name: networkgateway-1
spec:
  selector: # service label selector
    istio: eastwest-gateway
  network: network-1 # optional
  cluster: cluster-1 # optional
  addresses: # optional
    - 200.0.0.1
    - eastwest-gateway.example.com
  port: 15443 # optional
```

**Pull request closed without merging**